canonical/maas-ui
# refactor: update machine lists on machine.updateNotify MAASENG-1860

## Done

- refactor: update machine lists on machine.updateNotify
  - update any displayed machine lists grouping based on the information passed by the `machine.updateNotify` 
  - create new `FetchGroupByKey` type narrowing down keys used for group by in the UI
  - remove `machine.list[callId].needsUpdate` and associated utilities
  - add missing `parent` property to `BaseMachine`
  - add original request params to each machine list instance in redux store
- drive-by improvements:
    - remove deprecated `@testing-library/react-hooks` and replace it with `@testing-library/react`

### Note
This change makes `useFetchMachinesWithGroupingUpdates` hook redundant. As a result it removes 2 additional API calls per received notification. 

Fetching for the complete machine count for a new group will be implemented as part of https://warthogs.atlassian.net/browse/MAASENG-1882

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to machine list page
- select group by architecture
- open a machine details page for one of the machines on the list in a new tab
- go to configuration tab and switch architecture to a different type
- make sure the machine has been moved to the correct group

---

- Go to machine list page
- select group by status
- select multiple machines with a "ready" status
- Click Troubleshooting -> Test
- make sure the machines have been moved to the correct group (Testing)

---

- Repeat the steps below for group by zone, architecture, resource pool, power status:
  - set the machine list to a new grouping value
  - open a machine from the list in a new tab and perform an action that will alter its grouping value either on the machine list directly or by going to machine details page
  - verify that the machine you performed the action on is displayed under the correct group after update
## Fixes

Fixes: https://warthogs.atlassian.net/browse/MAASENG-1860

## Launchpad issue

Related Launchpad maas issue in the form `lp#number`.

## Backports

In general, please propose fixes against _main_ rather than release branches (e.g. 2.7), unless the fix is only applicable for that specific release. Please apply backport labels to the PR (e.g. "Backport 2.7") for the appropriate releases to target.

Only bug and security fixes should be backported, new features should only land in main.

## Screenshots

It could be helpful to provide some screenshots to aid in QAing the change.
